### ISSUE TITLE
Some packages are not loading

### ISSUE THREAD

**Issue by bondzula** [hlissner/doom-emacs#286]
First a disclaimer, I'm new to emacs, so maybe I'm doing something wrong.
Anyhow, when i start my emacs, some packages are not loaded, such as counsel. so by trying `M-x` i get 

`Wrong type argument: commandp, counsel-M-x` 

From what i could understand, you are lazy loading most of the packages, and counsel simply isn't loaded, if i change `:after ivy` to simply `:demand t` it works just fine. But i didn't want to lose lazy loading and what i did is use `:commands (counsel-find-file counsel-M-x etc...)` and that worked for me. But the same thing happen with `company-anaconda` where i wouldn't get any language completion, only words from the buffer where i typed in, and if i change `:after ...` to `:demand t` it works just fine.
So I'm wondering, is this a bug? maybe misuse on my part? should i just ditch lazy loading and go for `:demand t` as that works just fine

I am using 
`GNU Emacs 25.3.1`
`Arch Linux 4.14.3`

and all i did was copy init.example.el to init.el, run : 
`make install`
`make update`
`make compile-core`

Tnx for all the hard work :)

**Comment by hlissner** [comment]
`:demand t` shouldn't be necessary, ivy is loaded immediately, therefore counsel should be loaded immediately after. That said, adding `:demand t` won't hurt either, if that's what got it working for you.

However, this could indicate that something else went wrong before counsel could load. Have you checked the `*Messages*` buffer to see if any errors crop up (`M-x view-echo-area-messages`)? Or `M-x list-buffers` to see if a `*Warnings*` buffer silently appeared?

**Comment by bondzula** [comment]
I just checked, there are no errors at all. If counsel was the only package affected i wouldn't mind having that on `:demand t` but i don't relay need company-anaconda to load every time as i won't be editing python all the time, and i didn't relay test this much, but chances are that more packages are affected, maybe I'm missing something here, but it sure seems like `:after ...` hook doesn't work :/

**Comment by hlissner** [comment]
Could you check to see if `use-package-keywords` contains the `:after` keyword (with `M=x describe-variable RET use-package-keywords`)?

Could you also try replacing the `:after ...` with `:requires ...`?

**Comment by bondzula** [comment]
`use-package-keywords` does contain `:after` and changing it to `:requires` makes no difference

**Comment by hlissner** [comment]
Turns out, as of recently, this is [expected behavior for use-package](https://github.com/jwiegley/use-package/issues/575).

You may use `:demand t` in that case. This has some farther-reaching implications for Doom that I will address more comprehensively by 2.0.8. I'll leave this open until it's fixed.

**Comment by mantlepro** [comment]
I can confirm that replacing `:after ivy` with `demand t` fixed it for me.

    # ~/.emacs.d/modules/completion/ivy/config.el
    
    (def-package! counsel
      :demand t
    ...

**Comment by bondzula** [comment]
After reading this https://github.com/jwiegley/use-package/issues/562 it seem that old approch worked becouse of the bug, all that needs to be done is add `:demad t` after `:after ....`, as it will load that package only after condition is met . So it still is lazy loading. I've tested this with `python-mode` and it worked, `anaconda-mode` and `company-anaconda` are loaded only after python mode

**Comment by hlissner** [comment]
I've made some big changes, mainly centered around setting `use-package-always-defer` back to its default (`nil`). Something I've meant to do for a while.

It is in `develop` while I test it, but should resolve your issue.

@bondzula I've found another solution to the problem, but thanks for explaining it to me. That makes sense.

I'll keep this issue open until the fix has be merged into `master`.

**Comment by hlissner** [comment]
This is now in master, so I'll consider this issue resolved. Cheers!
